### PR TITLE
{data}[foss/2019a] ITK v5.0.1 w/ Python 3.7.2

### PR DESCRIPTION
**Pull request by murfreesboro**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by murfreesboro** [comment]
Test report by @murfreesboro
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
dev806.vampire - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU           E5620  @ 2.40GHz, Python 2.7.5
See https://gist.github.com/e984145aba713de29f343ff586d3a3db for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 0 out of 2 (1 easyconfigs in this PR)
login2.cerberus.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6126 CPU @ 2.60GHz, Python 2.7.5
See https://gist.github.com/6f2ce15edb8a3f31be72248f1d57dead for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 1 out of 2 (1 easyconfigs in this PR)
login2.cerberus.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6126 CPU @ 2.60GHz, Python 2.7.5
See https://gist.github.com/602c73dcec34a9dc8ad65c0c2c6d2a56 for a full test report.

**Comment by lexming** [comment]
```
build/ITK/5.0.1/foss-2019a-Python-3.7.2/ITK-5.0.1/Modules/ThirdParty/Eigen3/src/itkeigen/Eigen/src/Core/arch/AVX/PacketMath.h:169:54:` error: invalid output size for constraint '+x'`
__asm__("vfmadd231ps %[a], %[b], %[c]" : [c] "+x" (res) : [a] "x" (a), [b] "x" (b));
```
Weird error... maybe it's not compatible with AVX512

**Comment by lexming** [comment]
I hit the same error by using an external `Eigen/3.3.7`
```
/apps/brussel/CO7/skylake/software/Eigen/3.3.7/include/Eigen/src/Core/arch/AVX/PacketMath.h:169:54: error: invalid output size for
      constraint '+x'
  __asm__("vfmadd231ps %[a], %[b], %[c]" : [c] "+x" (res) : [a] "x" (a), [b] "x" (b));
```

**Comment by murfreesboro** [comment]
is that related to the AVX structure? Maybe you used an old CPU that does not support the AVX @lexming ?

**Comment by lexming** [comment]
Yes, it's definitely related to AVX. That's why I was trying to use the installation of Eigen that we already have working in our cluster. The funny part is that I'm using a skylake-x with a rather recent AVX.
After some tests with `-march`, I found that the error is (surprisingly) caused by a `-march=native` inserted by EB in the cmake configuration (in `CMAKE_C_FLAGS` and `MAKE_CXX_FLAGS`). CMake sets `ITK_CXX_OPTIMIZATION_FLAGS="-mtune=native -march=corei7"` and that actually works well, but with `-march=native` fails.

**Comment by murfreesboro** [comment]
Thank you @lexming , do you think I should disable the -march=native used by EB?

**Comment by lexming** [comment]
I got it to build with `toolchainopts = {'pic': True, 'cstd': 'c++11', 'optarch': False}`

**Comment by lexming** [comment]
We will also need to have shared objects, please add `-DBUILD_SHARED_LIBS=ON` to `configopts`.

**Comment by lexming** [comment]
One last suggestion, ITK is currently building several bundled *third party* libraries instead of using the dependencies of EB. For instance, HDF5, ZLIB, JPEG, PNG and TIFF. I got a working build with external dependencies at https://github.com/lexming/easybuild-easyconfigs/blob/itk/easybuild/easyconfigs/i/ITK/ITK-5.0.1-foss-2019a-Python-3.7.2.eb

**Comment by murfreesboro** [comment]
Hi @lexming your eb file looks very good. I think your eb file is much better than mine. Therefore if you prefer I am glad to close this PR and you can submit your PR. Or would you mind that we can submit your eb file in this PR to replace mine? Either way is good to me, so please just let me know.

**Comment by lexming** [comment]
I do not mind if you update this PR with my easyconfig for `ITK-5.0.1`, please do.

**Comment by murfreesboro** [comment]
Hi @lexming would you mind to give me your email address? I have put your name and mine on top as authors so that when people has problem with this eb file they can reach out to us.

**Comment by lexming** [comment]
Sure, please use alex.domingo.toro@vub.be

**Comment by murfreesboro** [comment]
Test report by @murfreesboro
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
dev806.vampire - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU           E5620  @ 2.40GHz, Python 2.7.5
See https://gist.github.com/564cd5e89a3c59df4aabf8edffcdfbbe for a full test report.

File: easybuild/easyconfigs/i/ITK/ITK-5.0.1-foss-2019a-Python-3.7.2.eb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+version = '5.0.1'
+versionsuffix = '-Python-%(pyver)s'
+
+homepage = 'https://itk.orgF'
+description = """Insight Segmentation and Registration Toolkit (ITK) provides
+ an extensive suite of software tools for registering and segmenting
+ multidimensional imaging data."""
+
+toolchain = {'name': 'foss', 'version': '2019a'}
+toolchainopts = {'pic': True, 'cstd': 'c++11', 'optarch': False}
```
[the remark this answers is not in the document]
Comment (murfreesboro) [reply easybuild/easyconfigs/i/ITK/ITK-5.0.1-foss-2019a-Python-3.7.2.eb]: Hi @boegel @lexming found that with `-march=native` he got errors in compilation, that's related to the eigen software. So we disable it. I will add a comment line in the eb file.

File: easybuild/easyconfigs/i/ITK/ITK-5.0.1-foss-2019a-Python-3.7.2.eb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+# Vrije Universiteit Brussel (VUB)
+# Vanderbilt University
+#
+easyblock = 'CMakeMake'
+
+name = 'ITK'
+version = '5.0.1'
+versionsuffix = '-Python-%(pyver)s'
+
+homepage = 'https://itk.orgF'
```
[the remark this answers is not in the document]
Comment (murfreesboro) [reply easybuild/easyconfigs/i/ITK/ITK-5.0.1-foss-2019a-Python-3.7.2.eb]: yeah, corrected it.

**Comment by boegel** [comment]
@lexming I think it's worth reporting that problem to `Eigen` (via http://eigen.tuxfamily.org/bz/) and getting some more feedback on it... Are you up for doing that?

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2455.golett.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/bd6b587674885db7f1e03402251c9716 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2066.delcatty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/d18768b7a7ebe48f0198b695546ab283 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3151.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/f364543c41b78d925699fc78f0720bd9 for a full test report.

**Comment by boegel** [comment]
Going in, thanks @murfreesboro!

**Comment by lexming** [comment]
> @lexming I think it's worth reporting that problem to `Eigen` (via http://eigen.tuxfamily.org/bz/) and getting some more feedback on it... Are you up for doing that?

@boegel This problem is not a fault in `Eigen`. It triggered with some headers of `Eigen` but the limitation is in `ITK` itself which does not support AVX512. `ITK` does its job and sets appropriate `march` and `mtune` directives (_e.g._ `ITK_CXX_OPTIMIZATION_FLAGS="-mtune=native -march=corei7"`). The problem arises when EB overrides the ITK settings with `-march=native` on systems with AVX512.

**Comment by boegel** [comment]
@lexming OK, thanks for clarifying!

**Pull request merged**